### ISSUE TITLE
Does not filter color and fontsize

### ISSUE THREAD

**Issue by bwl21** [jinhofer/pastespecial#7]
This is one of the most urgently required plugins. In particular while working with Chrome, cut&paste in the editor adds a lot of style garbage. So I highly appreciate this plugin (even if I think that ATTO should cleanup the pasted HTML by itself zwinkernd

So I installed it on my test drive. It seems to pass some styles per default:

I made an experiment in Chrome:

1. create a simple paragraph
2. do cut & paste
3. paste from word
4. paste from other.

The raw HTML looks like this. You can see that direct paste adds a lot of styles. Add "from word" and "add from other" yields the same result. I added a style named "foo" to the configuration of "paste from other", just to ensure that no hidden default (e.g. the one from word) is applied. But color and font size are always pasted.

```html
<p>paragraph with <b>bold .</b>
</p>
<p></p>
<p>
  <br>
</p>
<p>-------------------- direct paste</p>
<p>
</p>
<p style="margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 20px; orphans: auto; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255);">Simple paragraph with&nbsp;<b>bold</b>&nbsp;.</p>
<br>-------------------- paste from other
<p></p>
<p>
</p>
<p style="color: rgb(51, 51, 51);font-size: 14px;">Simple paragraph with&nbsp;<b>bold</b>&nbsp;.</p>

<p>--------------------- paste from word</p>
<p></p>
<p style="color: rgb(51, 51, 51);font-size: 14px;">Simple paragraph with&nbsp;<b>bold</b>&nbsp;.</p>
<p style="color: rgb(51, 51, 51);font-size: 14px;">
  <br>
</p>
<p></p>
<p></p>
<p></p>
````

I still find ATTO being indeed simpler to use than TinyMce but, the generated HTML is still not optimal (not to say a nightmare). You can see for example that the paste does som ugly things ( e.g. add `&nbsp;`around inline tags.

This is also the comment on https://moodle.org/plugins/atto_pastespecial.

**Comment by jinhofer** [comment]
I see what you mean. The reason color and fontsize always show up is that color and fontsize are always a part of whatever is pasted. Text editors such as Word and Google Documents don't function like HTML where the font-size is set by page and only specified when different; it's always specified in every line of those documents. Therefore, I can't justify removing color and font-size as they can are always present but can be different.

As for the &nbsp; included, these were originally included from the source (if you see the end of the direct paste):
```
...rgb(255, 255, 255);">Simple paragraph with&nbsp;<b>bold</b>&nbsp;.</p>
```

If you want color and font-size removed, you can use "Paste unformatted".

**Comment by bwl21** [comment]
Even if fontsize and color is always present, there is no reason to always copy them to the html in ATTO. If one wants this, then he can configure paste-special accordingly. Is there really a need to hardcode this if there is an administrator who knows what he is doing? I have a great hope in paste-special to be the vehicle to cleanup the code according tot the strategy of the Moodle site. I would highly appreciate if you could give the users the chance to do this in consequence.

Paste unformatted also removes the HTML tags and thereby fully destroys the structure. It even converts inline elements to extra lines. Paste unformatted should retain paragraphs, but remove inline tags while keeping the content inline.

**Comment by jinhofer** [comment]
Actually, you can prevent fontsize and color from being brought in (if I properly understand you suggestion)! Simply remove 'font-size' and/or 'color' from /admin/settings.php?section=atto_pastespecial_settings
There are other CSS properties that you can add/remove from there as you see fit.

So I see. It handles inline span easily enough, but not bold tags. I will work on this as well.

**Comment by bwl21** [comment]
As a pity it does not work. Here is the configuration which Yields the "other" result shown in the opening message of this issue.

it seems that paste_special cannot filter 'font-size' or 'color'

![screenshot_426](https://cloud.githubusercontent.com/assets/2022976/11817933/a3351ca6-a357-11e5-8c21-f6dea1f047fd.jpg)

**Comment by jinhofer** [comment]
I seem to be unable to reproduce this issue. I set "Other" to "foo" as you did, I open pastespecial, and I make the inner HTML of the editablecontentdiv:
```
<p style="margin: 0px 0px 10px; color: rgb(51, 51, 51); font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 20px; orphans: auto; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255);">Simple paragraph with&nbsp;<b>bold</b>&nbsp;.</p>
```
Then when I paste, I get:
```
<p>Simple paragraph with&nbsp;<b>bold</b>&nbsp;.</p><p><br></p>
```

Are you seeing a different HTML when you paste?

**Comment by bwl21** [comment]
I tried again and found out that there are some side effects. It appears to me that:

1. if you clean all text from the editor, atto keeps a pargraph with some styles take from whatever ...
2. Atto always pastes into a pargraph at least into an HTML - element. In other words, atto cannot insert something between to HTML elements.
3. Atto promotes the styles of the paragraph, where you paste into to the pasted elements.

In other words, it seems to be nearly impossible to create clean HTML with ATTO except by applying "clean format".

I have no Idea what we could do to make things more intuitive. But as the style filter works propoerly maybe we close this ticket and open a new one with requests for fixes of the structural deficiencies.

**Comment by jinhofer** [comment]
We can repurpose this ticket for those. Would you be able to clarify on those 3 points, and maybe provide brief examples so I know for sure what you're talking about?

**Comment by bwl21** [comment]
I wanted to assign the ticket to me but it seems that I can't. Could you do this. I then will try to clarify requirements with respect to structural issues.

**Comment by jinhofer** [comment]
It doesn't seem like I can assign this issue to you either. If you can clarify the requirements here, I can certainly work on them!

**Comment by bwl21** [comment]
I am afraid, we need to perform some structural manipulations before pasting. There is a way to perform xslt - transformations in modern browsers.

I will work offline on requirements and post theme here later. Alternatively we could turn on the Wiki and I can work on it online. In this case, The top level page can be kept clean by referring to a requirements page.

I basically think of 

1. find the appropriate insertion point:
    * block-level elements (table, div, paragraph) shall be pasted after the current paragraph
    * inline elments are pasted in the current paragraph
2. cleanup lists - Word is really weird with that
3. cleanup tables - I think pasting tables is really important

**Comment by bwl21** [comment]
I would like to play around with an xslt script to do some structurual manipulations. Could you give me a hint if there is another way to debug than to install the plugin in a Moodle-Instance. In particular, I have to uninstall the plugin before I reinstall it.

Would it help if I create a demonstrator as HTML - page with embedded Javascript which does the transformation.

**Comment by jinhofer** [comment]
If I understand what you're asking about (I'm foreign to xslt), you can do some inline debugging in Chrome, but that resets whenever you reload the page. The best bet may be to create the demonstrator, then we can look at how to integrate that.

**Comment by bwl21** [comment]
Just a note ... It was pushed the stack, but I still have it on the radar. I even started to build a playground independent from Moodle. But I even suffer from the fact that Word sometimes yields non wellformed xml.
